File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemo/Storage/AutoPayStorage.swift
//
//  AutoPayStorage.swift
//  PaykitDemo
//
//  Persistent storage for auto-pay settings using Keychain.
//

import Foundation

/// Auto-pay global settings
struct AutoPaySettings: Codable {
    var isEnabled: Bool
    var globalDailyLimitSats: Int64
    var currentDailySpentSats: Int64
    var lastResetDate: Date
    var requireConfirmation: Bool
    var notifyOnPayment: Bool
    
    init() {
        self.isEnabled = false
        self.globalDailyLimitSats = 100000 // 100k sats default
        self.currentDailySpentSats = 0
        self.lastResetDate = Date()
        self.requireConfirmation = false
        self.notifyOnPayment = true
    }
    
    mutating func resetIfNeeded() {
        let calendar = Calendar.current
        if !calendar.isDateInToday(lastResetDate) {
            currentDailySpentSats = 0
            lastResetDate = Date()
        }
    }
    
    var remainingDailyLimitSats: Int64 {
        max(0, globalDailyLimitSats - currentDailySpentSats)
    }
    
    var dailyUsagePercent: Double {
        guard globalDailyLimitSats > 0 else { return 0 }
        return Double(currentDailySpentSats) / Double(globalDailyLimitSats) * 100
    }
}

/// A peer-specific spending limit
struct PeerSpendingLimit: Identifiable, Codable {
    let id: String
    var peerPubkey: String
    var peerName: String
    var limitSats: Int64
    var spentSats: Int64
    var period: String  // daily, weekly, monthly
    var lastResetDate: Date
    
    init(peerPubkey: String, peerName: String, limitSats: Int64, period: String = "daily") {
        self.id = peerPubkey
        self.peerPubkey = peerPubkey
        self.peerName = peerName
        self.limitSats = limitSats
        self.spentSats = 0
        self.period = period
        self.lastResetDate = Date()
    }
    
    mutating func resetIfNeeded() {
        let calendar = Calendar.current
        let shouldReset: Bool
        
        switch period.lowercased() {
        case "daily":
            shouldReset = !calendar.isDateInToday(lastResetDate)
        case "weekly":
            shouldReset = !calendar.isDate(lastResetDate, equalTo: Date(), toGranularity: .weekOfYear)
        case "monthly":
            shouldReset = !calendar.isDate(lastResetDate, equalTo: Date(), toGranularity: .month)
        default:
            shouldReset = false
        }
        
        if shouldReset {
            spentSats = 0
            lastResetDate = Date()
        }
    }
    
    var remainingSats: Int64 {
        max(0, limitSats - spentSats)
    }
    
    var usagePercent: Double {
        guard limitSats > 0 else { return 0 }
        return Double(spentSats) / Double(limitSats) * 100
    }
}

/// An auto-pay rule
struct AutoPayRule: Identifiable, Codable {
    let id: String
    var name: String
    var isEnabled: Bool
    var maxAmountSats: Int64?
    var allowedMethods: [String]
    var allowedPeers: [String]  // Empty = all peers
    var requireConfirmation: Bool
    var createdAt: Date
    
    init(
        name: String,
        maxAmountSats: Int64? = nil,
        allowedMethods: [String] = [],
        allowedPeers: [String] = []
    ) {
        self.id = UUID().uuidString
        self.name = name
        self.isEnabled = true
        self.maxAmountSats = maxAmountSats
        self.allowedMethods = allowedMethods
        self.allowedPeers = allowedPeers
        self.requireConfirmation = false
        self.createdAt = Date()
    }
    
    func matches(amount: Int64, method: String, peer: String) -> Bool {
        guard isEnabled else { return false }
        
        // Check amount
        if let max = maxAmountSats, amount > max {
            return false
        }
        
        // Check method
        if !allowedMethods.isEmpty && !allowedMethods.contains(method) {
            return false
        }
        
        // Check peer
        if !allowedPeers.isEmpty && !allowedPeers.contains(peer) {
            return false
        }
        
        return true
    }
}

/// Manages persistent storage of auto-pay settings
class AutoPayStorage {
    
    private let keychain: KeychainStorage
    private let settingsKey = "paykit.autopay.settings"
    private let limitsKey = "paykit.autopay.limits"
    private let rulesKey = "paykit.autopay.rules"
    
    // In-memory cache
    private var settingsCache: AutoPaySettings?
    private var limitsCache: [PeerSpendingLimit]?
    private var rulesCache: [AutoPayRule]?
    
    init(keychain: KeychainStorage = KeychainStorage(serviceIdentifier: "com.paykit.demo")) {
        self.keychain = keychain
    }
    
    // MARK: - Settings
    
    func getSettings() -> AutoPaySettings {
        if var cached = settingsCache {
            cached.resetIfNeeded()
            return cached
        }
        
        do {
            guard let data = try keychain.retrieve(key: settingsKey) else {
                return AutoPaySettings()
            }
            var settings = try JSONDecoder().decode(AutoPaySettings.self, from: data)
            settings.resetIfNeeded()
            settingsCache = settings
            return settings
        } catch {
            print("AutoPayStorage: Failed to load settings: \(error)")
            return AutoPaySettings()
        }
    }
    
    func saveSettings(_ settings: AutoPaySettings) throws {
        let data = try JSONEncoder().encode(settings)
        try keychain.store(key: settingsKey, data: data)
        settingsCache = settings
    }
    
    // MARK: - Peer Limits
    
    func getPeerLimits() -> [PeerSpendingLimit] {
        if var cached = limitsCache {
            for i in cached.indices {
                cached[i].resetIfNeeded()
            }
            return cached
        }
        
        do {
            guard let data = try keychain.retrieve(key: limitsKey) else {
                return []
            }
            var limits = try JSONDecoder().decode([PeerSpendingLimit].self, from: data)
            for i in limits.indices {
                limits[i].resetIfNeeded()
            }
            limitsCache = limits
            return limits
        } catch {
            print("AutoPayStorage: Failed to load limits: \(error)")
            return []
        }
    }
    
    func savePeerLimit(_ limit: PeerSpendingLimit) throws {
        var limits = getPeerLimits()
        if let index = limits.firstIndex(where: { $0.id == limit.id }) {
            limits[index] = limit
        } else {
            limits.append(limit)
        }
        try persistLimits(limits)
    }
    
    func deletePeerLimit(id: String) throws {
        var limits = getPeerLimits()
        limits.removeAll { $0.id == id }
        try persistLimits(limits)
    }
    
    // MARK: - Rules
    
    func getRules() -> [AutoPayRule] {
        if let cached = rulesCache {
            return cached
        }
        
        do {
            guard let data = try keychain.retrieve(key: rulesKey) else {
                return []
            }
            let rules = try JSONDecoder().decode([AutoPayRule].self, from: data)
            rulesCache = rules
            return rules
        } catch {
            print("AutoPayStorage: Failed to load rules: \(error)")
            return []
        }
    }
    
    func saveRule(_ rule: AutoPayRule) throws {
        var rules = getRules()
        if let index = rules.firstIndex(where: { $0.id == rule.id }) {
            rules[index] = rule
        } else {
            rules.append(rule)
        }
        try persistRules(rules)
    }
    
    func deleteRule(id: String) throws {
        var rules = getRules()
        rules.removeAll { $0.id == id }
        try persistRules(rules)
    }
    
    // MARK: - Private
    
    private func persistLimits(_ limits: [PeerSpendingLimit]) throws {
        let data = try JSONEncoder().encode(limits)
        try keychain.store(key: limitsKey, data: data)
        limitsCache = limits
    }
    
    private func persistRules(_ rules: [AutoPayRule]) throws {
        let data = try JSONEncoder().encode(rules)
        try keychain.store(key: rulesKey, data: data)
        rulesCache = rules
    }
}

File: paykit-mobile/android-demo/app/src/main/java/com/paykit/demo/storage/AutoPayStorage.kt
package com.paykit.demo.storage

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.paykit.demo.model.AutoPayRule
import com.paykit.demo.model.AutoPaySettings
import com.paykit.demo.model.PeerSpendingLimit
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Manages persistent storage of auto-pay settings using EncryptedSharedPreferences.
 */
class AutoPayStorage(context: Context) {
    
    companion object {
        private const val TAG = "AutoPayStorage"
        private const val PREFS_NAME = "paykit_autopay"
        private const val SETTINGS_KEY = "autopay_settings"
        private const val LIMITS_KEY = "autopay_limits"
        private const val RULES_KEY = "autopay_rules"
    }
    
    private val json = Json { 
        ignoreUnknownKeys = true 
        encodeDefaults = true
    }
    
    private val prefs by lazy {
        try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            
            EncryptedSharedPreferences.create(
                context,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create encrypted prefs, falling back", e)
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        }
    }
    
    private var settingsCache: AutoPaySettings? = null
    private var limitsCache: List<PeerSpendingLimit>? = null
    private var rulesCache: List<AutoPayRule>? = null
    
    // MARK: - Settings
    
    fun getSettings(): AutoPaySettings {
        settingsCache?.let { return it.resetIfNeeded() }
        
        return try {
            val jsonString = prefs.getString(SETTINGS_KEY, null) ?: return AutoPaySettings()
            val settings = json.decodeFromString<AutoPaySettings>(jsonString).resetIfNeeded()
            settingsCache = settings
            settings
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load settings: ${e.message}")
            AutoPaySettings()
        }
    }
    
    fun saveSettings(settings: AutoPaySettings) {
        try {
            val jsonString = json.encodeToString(settings)
            prefs.edit().putString(SETTINGS_KEY, jsonString).apply()
            settingsCache = settings
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save settings: ${e.message}")
        }
    }
    
    // MARK: - Peer Limits
    
    fun getPeerLimits(): List<PeerSpendingLimit> {
        limitsCache?.let { cached ->
            return cached.map { limit -> limit.resetIfNeeded() }
        }
        
        return try {
            val jsonString = prefs.getString(LIMITS_KEY, null) ?: return emptyList()
            val decoded: List<PeerSpendingLimit> = json.decodeFromString(jsonString)
            val limits: List<PeerSpendingLimit> = decoded.map { limit -> limit.resetIfNeeded() }
            limitsCache = limits
            limits
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load limits: ${e.message}")
            emptyList()
        }
    }
    
    fun savePeerLimit(limit: PeerSpendingLimit) {
        val limits = getPeerLimits().toMutableList()
        val index = limits.indexOfFirst { it.id == limit.id }
        
        if (index >= 0) {
            limits[index] = limit
        } else {
            limits.add(limit)
        }
        
        persistLimits(limits)
    }
    
    fun deletePeerLimit(id: String) {
        val limits = getPeerLimits().toMutableList()
        limits.removeAll { it.id == id }
        persistLimits(limits)
    }
    
    // MARK: - Rules
    
    fun getRules(): List<AutoPayRule> {
        rulesCache?.let { return it }
        
        return try {
            val jsonString = prefs.getString(RULES_KEY, null) ?: return emptyList()
            val rules = json.decodeFromString<List<AutoPayRule>>(jsonString)
            rulesCache = rules
            rules
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load rules: ${e.message}")
            emptyList()
        }
    }
    
    fun saveRule(rule: AutoPayRule) {
        val rules = getRules().toMutableList()
        val index = rules.indexOfFirst { it.id == rule.id }
        
        if (index >= 0) {
            rules[index] = rule
        } else {
            rules.add(rule)
        }
        
        persistRules(rules)
    }
    
    fun deleteRule(id: String) {
        val rules = getRules().toMutableList()
        rules.removeAll { it.id == id }
        persistRules(rules)
    }
    
    // MARK: - Private
    
    private fun persistLimits(limits: List<PeerSpendingLimit>) {
        try {
            val jsonString = json.encodeToString(limits)
            prefs.edit().putString(LIMITS_KEY, jsonString).apply()
            limitsCache = limits
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save limits: ${e.message}")
        }
    }
    
    private fun persistRules(rules: List<AutoPayRule>) {
        try {
            val jsonString = json.encodeToString(rules)
            prefs.edit().putString(RULES_KEY, jsonString).apply()
            rulesCache = rules
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save rules: ${e.message}")
        }
    }
}

